File: forms/employee-offboarding-checklist/full-stack-with-loco-tera-htmx-alpine/tests/engine/checklist_validator_test.rs
use employee_offboarding_checklist_tera_crate::engine::checklist_validator::{run_rules, validate};
use employee_offboarding_checklist_tera_crate::engine::types::*;
use employee_offboarding_checklist_tera_crate::engine::validation_rules::all_rules;

/// Build a fully-completed offboarding assessment where every rule is
/// satisfied.
fn fully_completed_case() -> AssessmentData {
    AssessmentData {
        employee_details: EmployeeDetails {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            employee_id: "EMP-001".to_string(),
            email: "jane.doe@example.org".to_string(),
            job_title: "Senior Nurse".to_string(),
            department: "Cardiology".to_string(),
            line_manager: "Mary Smith".to_string(),
            start_date: "2018-01-15".to_string(),
            last_working_day: "2026-05-31".to_string(),
            reason_for_leaving: "resignation".to_string(),
            reason_for_leaving_other: String::new(),
            employee_details_notes: String::new(),
        },
        exit_interview: ExitInterview {
            interview_offered: "yes".to_string(),
            interview_completed: "yes".to_string(),
            interview_date: "2026-05-28".to_string(),
            interviewer_name: "HR Officer".to_string(),
            feedback_provided: "yes".to_string(),
            feedback_documented: "yes".to_string(),
            concerns_raised: "no".to_string(),
            concerns_details: String::new(),
            exit_interview_notes: String::new(),
        },
        knowledge_transfer: KnowledgeTransfer {
            successor_identified: "yes".to_string(),
            successor_name: "John Apprentice".to_string(),
            handover_document_created: "yes".to_string(),
            handover_meetings_held: "yes".to_string(),
            work_in_progress_documented: "yes".to_string(),
            key_contacts_shared: "yes".to_string(),
            sop_files_transferred: "yes".to_string(),
            clinical_caseload_reassigned: "yes".to_string(),
            knowledge_transfer_notes: String::new(),
        },
        equipment_return: EquipmentReturn {
            laptop_returned: "yes".to_string(),
            laptop_asset_tag: "LAP-100".to_string(),
            mobile_phone_returned: "yes".to_string(),
            id_badge_returned: "yes".to_string(),
            keys_returned: "yes".to_string(),
            uniform_returned: "yes".to_string(),
            parking_pass_returned: "yes".to_string(),
            other_equipment_returned: "na".to_string(),
            other_equipment_description: String::new(),
            equipment_return_notes: String::new(),
        },
        access_revocation: AccessRevocation {
            email_revoked: "yes".to_string(),
            ehr_epr_revoked: "yes".to_string(),
            vpn_revoked: "yes".to_string(),
            active_directory_disabled: "yes".to_string(),
            building_access_revoked: "yes".to_string(),
            cloud_apps_revoked: "yes".to_string(),
            smartcard_deactivated: "yes".to_string(),
            data_download_audit_performed: "yes".to_string(),
            unauthorised_download_detected: "no".to_string(),
            access_revocation_notes: String::new(),
        },
        final_payroll_benefits: FinalPayrollBenefits {
            final_salary_calculated: "yes".to_string(),
            accrued_holiday_paid: "yes".to_string(),
            expenses_reimbursed: "yes".to_string(),
            pension_information_provided: "yes".to_string(),
            p45_issued: "yes".to_string(),
            benefits_terminated: "yes".to_string(),
            payroll_details_confirmed: "yes".to_string(),
            final_payment_date: "2026-06-15".to_string(),
            final_payroll_notes: String::new(),
        },
        references_recommendations: ReferencesRecommendations {
            reference_consent_given: "yes".to_string(),
            reference_contact_details_recorded: "yes".to_string(),
            recommendation_letter_requested: "no".to_string(),
            recommendation_letter_provided: "no".to_string(),
            reference_policy_explained: "yes".to_string(),
            references_notes: String::new(),
        },
        non_disclosure_post_employment: NonDisclosurePostEmployment {
            confidentiality_reaffirmed: "yes".to_string(),
            nda_signed: "yes".to_string(),
            restrictive_covenants_explained: "yes".to_string(),
            restrictive_covenants_acknowledged: "yes".to_string(),
            intellectual_property_assigned: "yes".to_string(),
            data_returned_or_destroyed: "yes".to_string(),
            post_employment_obligations_explained: "yes".to_string(),
            nda_notes: String::new(),
        },
        forwarding_details: ForwardingDetails {
            forwarding_address_line1: "1 The Avenue".to_string(),
            forwarding_address_line2: String::new(),
            forwarding_city: "Bristol".to_string(),
            forwarding_postcode: "BS1 1AA".to_string(),
            forwarding_country: "United Kingdom".to_string(),
            personal_email: "jane@personal.example".to_string(),
            personal_phone: "07700 900000".to_string(),
            forwarding_details_confirmed: "yes".to_string(),
            forwarding_notes: String::new(),
        },
        signoff: Signoff {
            hr_signed_off: "yes".to_string(),
            hr_sign_off_name: "HR Officer".to_string(),
            hr_sign_off_date: "2026-05-31".to_string(),
            line_manager_signed_off: "yes".to_string(),
            line_manager_sign_off_name: "Mary Smith".to_string(),
            line_manager_sign_off_date: "2026-05-31".to_string(),
            it_signed_off: "yes".to_string(),
            it_sign_off_name: "IT Manager".to_string(),
            it_sign_off_date: "2026-05-31".to_string(),
            employee_acknowledged: "yes".to_string(),
            employee_sign_off_date: "2026-05-31".to_string(),
            signoff_notes: String::new(),
        },
    }
}

#[test]
fn empty_case_is_incomplete() {
    let data = AssessmentData::default();
    let result = validate(&data);
    assert_eq!(result.outcome, "incomplete");
    assert!(!result.blockers.is_empty());
}

#[test]
fn fully_completed_case_is_complete() {
    let data = fully_completed_case();
    let result = validate(&data);
    assert_eq!(result.outcome, "complete");
    assert_eq!(result.completion_percent, 100.0);
    assert!(result.blockers.is_empty());
    assert!(result.fired_rules.is_empty());
}

#[test]
fn missing_laptop_yields_incomplete_with_blocker() {
    let mut data = fully_completed_case();
    data.equipment_return.laptop_returned = String::new();
    let result = validate(&data);
    assert_eq!(result.outcome, "incomplete");
    assert!(result.blockers.iter().any(|b| b.id == "EQ-001"));
}

#[test]
fn missing_non_mandatory_yields_partial() {
    let mut data = fully_completed_case();
    // SO-004 (employee_acknowledged) is non-mandatory; firing it should
    // demote outcome from complete to partial.
    data.signoff.employee_acknowledged = String::new();
    let result = validate(&data);
    assert_eq!(result.outcome, "partial");
    assert!(result.fired_rules.iter().any(|r| r.id == "SO-004"));
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn run_rules_returns_only_fired() {
    let data = fully_completed_case();
    let fired = run_rules(&data);
    assert!(fired.is_empty(), "fully completed case should have no fired rules");
}

File: forms/employee-offboarding-checklist/front-end-with-svelte/src/lib/engine/checklist-validator.test.ts
import { describe, it, expect } from 'vitest';
import { validateChecklist } from './checklist-validator';
import { detectAdditionalFlags } from './flagged-issues';
import { validationRules } from './validation-rules';
import type { AssessmentData } from './types';

/**
 * A blank checklist with every field at its unanswered default. Kept inline
 * (rather than importing the store factory) so the test does not pull in
 * SvelteKit's `$app/environment` module, which is unavailable under Vitest.
 */
function createBlankChecklist(): AssessmentData {
	return {
		employeeDetails: {
			firstName: '',
			lastName: '',
			employeeId: '',
			email: '',
			jobTitle: '',
			department: '',
			lineManager: '',
			startDate: '',
			lastWorkingDay: '',
			reasonForLeaving: '',
			reasonForLeavingOther: '',
			employeeDetailsNotes: ''
		},
		exitInterview: {
			interviewOffered: '',
			interviewCompleted: '',
			interviewDate: '',
			interviewerName: '',
			feedbackProvided: '',
			feedbackDocumented: '',
			concernsRaised: '',
			concernsDetails: '',
			exitInterviewNotes: ''
		},
		knowledgeTransfer: {
			successorIdentified: '',
			successorName: '',
			handoverDocumentCreated: '',
			handoverMeetingsHeld: '',
			workInProgressDocumented: '',
			keyContactsShared: '',
			sopFilesTransferred: '',
			clinicalCaseloadReassigned: '',
			knowledgeTransferNotes: ''
		},
		equipmentReturn: {
			laptopReturned: '',
			laptopAssetTag: '',
			mobilePhoneReturned: '',
			idBadgeReturned: '',
			keysReturned: '',
			uniformReturned: '',
			parkingPassReturned: '',
			otherEquipmentReturned: '',
			otherEquipmentDescription: '',
			equipmentReturnNotes: ''
		},
		accessRevocation: {
			emailRevoked: '',
			ehrEprRevoked: '',
			vpnRevoked: '',
			activeDirectoryDisabled: '',
			buildingAccessRevoked: '',
			cloudAppsRevoked: '',
			smartcardDeactivated: '',
			dataDownloadAuditPerformed: '',
			unauthorisedDownloadDetected: '',
			accessRevocationNotes: ''
		},
		finalPayrollBenefits: {
			finalSalaryCalculated: '',
			accruedHolidayPaid: '',
			expensesReimbursed: '',
			pensionInformationProvided: '',
			p45Issued: '',
			benefitsTerminated: '',
			payrollDetailsConfirmed: '',
			finalPaymentDate: '',
			finalPayrollNotes: ''
		},
		referencesRecommendations: {
			referenceConsentGiven: '',
			referenceContactDetailsRecorded: '',
			recommendationLetterRequested: '',
			recommendationLetterProvided: '',
			referencePolicyExplained: '',
			referencesNotes: ''
		},
		nonDisclosurePostEmployment: {
			confidentialityReaffirmed: '',
			ndaSigned: '',
			restrictiveCovenantsExplained: '',
			restrictiveCovenantsAcknowledged: '',
			intellectualPropertyAssigned: '',
			dataReturnedOrDestroyed: '',
			postEmploymentObligationsExplained: '',
			ndaNotes: ''
		},
		forwardingDetails: {
			forwardingAddressLine1: '',
			forwardingAddressLine2: '',
			forwardingCity: '',
			forwardingPostcode: '',
			forwardingCountry: '',
			personalEmail: '',
			personalPhone: '',
			forwardingDetailsConfirmed: '',
			forwardingNotes: ''
		},
		signoff: {
			hrSignedOff: '',
			hrSignOffName: '',
			hrSignOffDate: '',
			lineManagerSignedOff: '',
			lineManagerSignOffName: '',
			lineManagerSignOffDate: '',
			itSignedOff: '',
			itSignOffName: '',
			itSignOffDate: '',
			employeeAcknowledged: '',
			employeeSignOffDate: '',
			signoffNotes: ''
		}
	};
}

/** A fully-completed checklist where every rule is satisfied. */
function createCompleteChecklist(): AssessmentData {
	const d = createBlankChecklist();
	d.employeeDetails = {
		...d.employeeDetails,
		firstName: 'Alex',
		lastName: 'Taylor',
		lastWorkingDay: '2026-07-31',
		reasonForLeaving: 'resignation'
	};
	d.exitInterview = {
		...d.exitInterview,
		interviewOffered: 'yes',
		interviewCompleted: 'yes',
		feedbackDocumented: 'yes'
	};
	d.knowledgeTransfer = {
		...d.knowledgeTransfer,
		successorIdentified: 'yes',
		handoverDocumentCreated: 'yes',
		workInProgressDocumented: 'yes',
		clinicalCaseloadReassigned: 'yes'
	};
	d.equipmentReturn = {
		...d.equipmentReturn,
		laptopReturned: 'yes',
		mobilePhoneReturned: 'yes',
		idBadgeReturned: 'yes',
		keysReturned: 'yes',
		uniformReturned: 'na',
		parkingPassReturned: 'na'
	};
	d.accessRevocation = {
		...d.accessRevocation,
		emailRevoked: 'yes',
		ehrEprRevoked: 'yes',
		vpnRevoked: 'yes',
		activeDirectoryDisabled: 'yes',
		buildingAccessRevoked: 'yes',
		cloudAppsRevoked: 'yes',
		dataDownloadAuditPerformed: 'yes',
		unauthorisedDownloadDetected: 'no'
	};
	d.finalPayrollBenefits = {
		...d.finalPayrollBenefits,
		finalSalaryCalculated: 'yes',
		accruedHolidayPaid: 'yes',
		p45Issued: 'yes',
		payrollDetailsConfirmed: 'yes'
	};
	d.referencesRecommendations = {
		...d.referencesRecommendations,
		referenceConsentGiven: 'yes',
		referencePolicyExplained: 'yes'
	};
	d.nonDisclosurePostEmployment = {
		...d.nonDisclosurePostEmployment,
		confidentialityReaffirmed: 'yes',
		restrictiveCovenantsExplained: 'yes',
		intellectualPropertyAssigned: 'yes',
		dataReturnedOrDestroyed: 'yes'
	};
	d.forwardingDetails = {
		...d.forwardingDetails,
		forwardingAddressLine1: '1 High Street',
		forwardingPostcode: 'AB1 2CD',
		forwardingDetailsConfirmed: 'yes'
	};
	d.signoff = {
		...d.signoff,
		hrSignedOff: 'yes',
		lineManagerSignedOff: 'yes',
		itSignedOff: 'yes',
		employeeAcknowledged: 'yes'
	};
	return d;
}

describe('Offboarding Completeness Validation', () => {
	it('returns Complete with 100% for a fully-completed checklist', () => {
		const result = validateChecklist(createCompleteChecklist());
		expect(result.outcome).toBe('complete');
		expect(result.completionPercent).toBe(100);
		expect(result.firedRules).toHaveLength(0);
		expect(result.blockers).toHaveLength(0);
		expect(result.additionalFlags).toHaveLength(0);
	});

	it('returns Partial when only non-blocking items are outstanding', () => {
		const d = createCompleteChecklist();
		// Non-mandatory, non-blocking item outstanding.
		d.referencesRecommendations.referencePolicyExplained = 'no';
		d.exitInterview.interviewCompleted = 'no';
		const result = validateChecklist(d);
		expect(result.outcome).toBe('partial');
		expect(result.blockers).toHaveLength(0);
		expect(result.firedRules.length).toBeGreaterThan(0);
	});

	it('returns Incomplete when a mandatory blocker is outstanding', () => {
		const d = createCompleteChecklist();
		d.accessRevocation.emailRevoked = 'no';
		const result = validateChecklist(d);
		expect(result.outcome).toBe('incomplete');
		expect(result.blockers.some((b) => b.id === 'AC-001')).toBe(true);
	});

	it('returns Incomplete for a blank checklist with completion below 100%', () => {
		const result = validateChecklist(createBlankChecklist());
		expect(result.outcome).toBe('incomplete');
		expect(result.completionPercent).toBeLessThan(100);
		expect(result.mandatoryTotal).toBeGreaterThan(0);
		expect(result.mandatorySatisfied).toBe(0);
	});

	it('has unique rule IDs', () => {
		const ids = validationRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Offboarding Flagged Issues Detection', () => {
	it('returns no flags for a fully-completed checklist', () => {
		const flags = detectAdditionalFlags(createCompleteChecklist());
		expect(flags).toHaveLength(0);
	});

	it('flags unauthorised data download as high priority', () => {
		const d = createCompleteChecklist();
		d.accessRevocation.unauthorisedDownloadDetected = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-AC-001' && f.priority === 'high')).toBe(true);
	});

	it('flags outstanding access channels', () => {
		const d = createCompleteChecklist();
		d.accessRevocation.vpnRevoked = 'no';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-AC-002')).toBe(true);
	});

	it('flags equipment not returned', () => {
		const d = createCompleteChecklist();
		d.equipmentReturn.laptopReturned = 'no';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-EQ-001')).toBe(true);
	});

	it('flags missing forwarding address as low priority', () => {
		const d = createCompleteChecklist();
		d.forwardingDetails.forwardingAddressLine1 = '';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-FWD-001' && f.priority === 'low')).toBe(true);
	});

	it('sorts flags by priority (high first)', () => {
		const flags = detectAdditionalFlags(createBlankChecklist());
		const order = { high: 0, medium: 1, low: 2 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
